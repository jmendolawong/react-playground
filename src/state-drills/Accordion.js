import React from 'react';

class Accordion extends React.Component {
    static defaultProps = {
        sections: []
    };

    state = {
        currentSection: null
    }

    handleClick = (index) => {
        this.setState ({
            currentSection: index
        })
    }

    renderListItems () {
        return this.props.sections.map((section, i) => (
            <li key={i}>
                <button type='button' onClick={()=>this.handleClick(i)}>
                    {section.title}
                </button>
                {this.state.currentSection === i && this.renderContent()}
            </li>
        ))
    }

    renderContent(){
        const newP = this.props.sections[this.state.currentSection]
        return (
            <p className='content'>{newP.content}</p>
        )
    }


    render(){
        return(
            <ul>
              {this.renderListItems()}  
            </ul>
        )
    }
}


export default Accordion