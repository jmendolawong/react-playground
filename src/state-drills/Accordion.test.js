import React from 'react';
import ReactDOM from 'react-dom'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json';
import Accordion from './Accordion';

describe('Accordion Component', () =>{
    const sectionsProp = [
        { title: 'Section 1',
          content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.'},
        { title: 'Section 2',
          content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!'},
        { title: 'Section 3',
          content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?'},
      ];

    it('renders without errors', ()=>{
        //create a div element
        //render the Accordion component in the newly created div
        //Unmount from div. All actions mean successful test
        const div = document.createElement('div');
        ReactDOM.render(<Accordion />, div);
        ReactDOM.unmountComponentAtNode(div);
    })
      
    it('renders empty li without a supplied prop', ()=>{
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('renders no sections as active by default', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('opens up a clicked section', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(0).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    
    it('opens up last clicked section after multiple clicks', ()=>{
        const wrapper = shallow(<Accordion sections={sectionsProp} />)
        wrapper.find('button').at(0).simulate('click')
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(2).simulate('click')
        wrapper.find('button').at(0).simulate('click')       
        expect(toJson(wrapper)).toMatchSnapshot()
    })
})