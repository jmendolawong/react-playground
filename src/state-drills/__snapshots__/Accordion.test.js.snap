// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component opens up a clicked section 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
    <p
      className="content"
    >
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component opens up last clicked section after multiple clicks 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
    <p
      className="content"
    >
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders empty li without a supplied prop 1`] = `<ul />`;

exports[`Accordion Component renders no sections as active by default 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders no sections by default 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      Section 3
    </button>
  </li>
</ul>
`;
